Breaking up Java quick-start, surface Java Signals and Queries (#431)
temporalio/documentation: docs/java/activities.md

sidebar_label: Activities
---

> ## What is an Activity?

Reviewer: Would it be ok to rewrite this paragraph some? If so let me know I can provide some suggestions

Author: Of course.

Reviewer: updated. thanks!

Having fault tolerant code that maintains state is useful.
But in most practical applications, a Workflow will be expected to act upon the external world.
Temporal supports such externally-facing code in the form of Activities.

An Activity is essentially a function that can execute any code like DB updates or service calls.
The Workflow is not allowed to directly call any external APIs; it can do this only through Activities.
The Workflow is essentially an orchestrator of Activities.
Activities that fail or timeout will be retried; therefore they should be written with care to be idempotent.

> ### Activity interface

Reviewer: Add space above?

Reviewer: added

Similar to Workflows, Activities in Temporal Java SDK programming model are classes which implement an Activity Interface:

```java
@ActivityInterface
public interface GreetingActivities {
    String composeGreeting(String greeting, String name);
}
```

An Activity interface must be annotated with the `@ActivityInterface`.
Each method that belongs to an `@ActivityInterface` defines a separate Activity type.
You can annotate each method in the Activity interface with the `@ActivityMethod` annotation, but this is completely optional.

```java
@ActivityInterface
public interface GreetingActivities {

>     @ActivityMethod(name=composeHello)

Reviewer: this is not needed

Reviewer: removed

    String composeGreeting(String greeting, String name);
}
```

Note that the `@ActivityMethod` annotation has a `name` parameter which can be used to define the Activity type.
If not specified, the method name (with the first letter capitalized) is used by default, so in the above example it would be `ComposeHello`. If the annotation did not specify the name then the Activity Type would be `ComposeGreeting`.

### Activity implementation

An Activity implementation is just a normal [POJO](https://en.wikipedia.org/wiki/Plain_old_Java_object).

> The `out` stream is passed as a parameter to the constructor to demonstrate that the Activity object can have any dependency.

Reviewer: This is because when registering an activity with worker you use the activity instance, and not the type like with Workflows. maybe we can add

Reviewer: added

sidebar_label: Activities
---

## What is an Activity?

Having fault tolerant code that maintains state is useful.
But in most practical applications, a Workflow will be expected to act upon the external world.
Temporal supports such externally-facing code in the form of Activities.

An Activity is essentially a function that can execute any code like DB updates or service calls.

> The Workflow is not allowed to directly call any external APIs;

Author: ```suggestion
The Workflow is not allowed to directly call any external APIs; it can do this only through Activities.
```

> It can do this only through Activities.

Reviewer 2: ```suggestion
it can do this only through Activities.
```

Author: Oh, ya my b.

> It can do this only through Activities.

Author: ```suggestion
```

The Workflow is essentially an orchestrator of Activities.

> 

Reviewer 2: ```suggestion
Activities that fail or timeout will be retried; therefore they should be written with care to be idempotent.
```

### Activity interface

Similar to Workflows, Activities in Temporal Java SDK programming model are classes which implement an Activity Interface:

```java
@ActivityInterface
public interface GreetingActivities {
    String composeGreeting(String greeting, String name);
}
```

An Activity interface must be annotated with the `@ActivityInterface`.

> Each method that belongs to an Activity interface defines a separate Activity type.

Reviewer 2: ```suggestion
Each method that belongs to an `@ActivityInterface` defines a separate Activity type.
```

You can annotate each method in the Activity interface with the `@ActivityMethod` annotation, but this is completely optional.

```java
@ActivityInterface
public interface GreetingActivities {
    @ActivityMethod(name=composeHello)
    String composeGreeting(String greeting, String name);
}
```

Note that the `@ActivityMethod` annotation has a `name` parameter which can be used to define the Activity type.

> If not specified, the method name (with the first letter capitalized) is used by default, so in the above example it would be  ComposeHello. If the annotation did not specify the name then the Activity Type would be "ComposeGreeting".

Reviewer 2: ```suggestion
If not specified, the method name (with the first letter capitalized) is used by default, so in the above example it would be `ComposeHello`. If the annotation did not specify the name then the Activity Type would be `ComposeGreeting`.
```

sidebar_label: Activities
---

## What is an Activity?

> Activities are implementations of certain tasks which need to be performed during workflow execution.

Author: ```suggestion
Activities are implementations of certain tasks which need to be performed during a Workflow execution.
```

Activity options allow you to specify different Activity timeout and retry options.

Calling a method on this interface invokes an Activity that implements this method.
An Activity invocation synchronously blocks until the Activity completes, fails, or times out.

> Even if Activity execution takes a few months, the Workflow code still sees it as a single synchronous invocation.

Author: ```suggestion
Even if an Activity's execution takes a few months, the Workflow code still sees it as a single synchronous invocation.
```
